_posts/2018-11-11-fs.md

```diff
@@ -0,0 +1,81 @@
+And thus, in case of the former, there is a greater probability of hitting a kernel bug, or exploiting a vulnerability.
+
+
+### Kernel Oops!
+We highlight this by a recently discovered bug in the ext4 filesystem code- [CVE-2018-10840](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2018-10840).
+In this case, removing the [extended attributes](http://man7.org/linux/man-pages/man5/attr.5.html) of a file on a crafted ext4 image, leads to a kernel oops due to a buffer overflow.
+The culprit syscall which allows a userspace program to trigger this bug is [`removexattr()`](http://man7.org/linux/man-pages/man2/removexattr.2.html).
+When testing with a Ubuntu 18.04<sup>2</sup> host system, we were able to trigger this bug in the kernel, from inside a Kata container.
+The following video shows a demo of this.
+
+[TODO: video]
+
+
+### The Bug
+We used the [original author's](https://bugzilla.kernel.org/show_bug.cgi?id=199347) crafted image to test this bug, which can be exposed with the following simple trigger code:
+
+```
+// mpoint is the path to the mounted crafted image
+static void activity(char *mpoint) {
+  char *xattr;
+  int err = asprintf(&xattr, "%s/foo/bar/xattr", mpoint);
+  char buf2[113];
+  memset(buf2, 0, sizeof(buf2));
+  listxattr(xattr, buf2, sizeof(buf2));
+  removexattr(xattr, "user.mime_type");
+}
+```
+The buffer overflow happens in the kernel's `fs/ext4/xattr.c:ext4_xattr_set_entry() 
+` function, because of a [missed check](https://elixir.bootlin.com/linux/v4.15/source/fs/ext4/xattr.c#L1671) on the `size` parameter of the [`memmove()`](http://man7.org/linux/man-pages/man3/memmove.3.html) function, which is negative when trying to remove an extended attribute (`user.mime_type`) for the file `foo/bar/xattr` in the crafted image. 
+The fix is fairly straightforward in this case--adding a [value check](https://bugzilla.kernel.org/attachment.cgi?id=276147&action=diff).
+
+
+
+### Nabla and GVisor
+Nabla's [philosophy](https://nabla-containers.github.io/) of limiting access to the host kernel means that a container's file access does not interface with the host's filesystem.
+Specifically, Nabla does not use a 9p or a passthrough protocol, with syscalls on the filesystem handled at the LibOS (rumprun) level.
+Thus, this bug would never get triggered<sup>3</sup>, even if the host kernel is vulnerable, since `removexattr()` is not propagated to the host. 
+Nabla's minimal *touch points* into the host can be seen [here](https://github.com/nabla-containers/solo5/blob/ukvm-linux-seccomp/ukvm/ukvm_hv_linux.c#L53-L106).
+
+The bug does not surface in case of Gvisor as well.
+Neither [Sentry](https://github.com/google/gvisor/blob/master/pkg/sentry/syscalls/linux/linux64.go#L239-L250) nor [Gofer](https://github.com/google/gvisor/blob/master/runsc/fsgofer/filter/config.go) support the syscall--"`Removexattr, requires filesystem support`"
```

Comment:
I would say it like: "The sentry does not support the syscall, and just in case, it's blocked by seccomp at the gofer side."